### PR TITLE
nixos/zapret: init

### PR DESCRIPTION
**Pull request by voronind-com**
## Things done

Implement a systemd service for NixOS that enables Zapret package system-wide.
I've been using it for a while now.

Compared to https://github.com/NixOS/nixpkgs/pull/327903 this module has whitelist/blacklist support as well as some other configurations. Also after running this for a while, I solved some pain-points like repetitive network slowdowns. And no "dirty" parts.

Also this is my first contribution. I'd really appreciate assistance with making things right this first time. Please tell me all your thoughts. Thanks!

Example configuration:
```nix
services.zapret = {
  enable = true;
  package = pkgs.zapret;
  params = [
    "--dpi-desync=fake,disorder2"
    "--dpi-desync-ttl=1"
    "--dpi-desync-autottl=2"
  ];
  whitelist = [
    "youtube.com"
    "googlevideo.com"
    "ytimg.com"
    "youtu.be"
  ];
  # blacklist = [ ];
  qnum = 200;
  configureFirewall = true;
  httpSupport = true;
};
```

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by voronind-com** [comment]
Looks like I actually need to create a release notes entry. On it now.

**Comment by voronind-com** [comment]
I totally missed and already open PR here: https://github.com/NixOS/nixpkgs/pull/327903

Closing this.

File: nixos/modules/services/networking/zapret.nix
Comment (kanashimia) [inline nixos/modules/services/networking/zapret.nix:123]: `/bin/kill`? On NixOS? 🙄

Comment (voronind-com) [reply nixos/modules/services/networking/zapret.nix:123]: Oopsie! Thank you very much

**Comment by ardishko** [comment]
Why not put this in a seperate repo under a flake? The maintainers said that the other PR was done dirtily and it will probably hang as an open PR for a while due to such.

**Comment by voronind-com** [comment]
> Why not put this in a seperate repo under a flake? The maintainers said that the other PR was done dirtily and it will probably hang as an open PR for a while due to such.

My implementation doesn't have the dirty part.

**Review by ardishko — APPROVED** [review]
LGTM. I hope this is merged into master soon. Russian and Turkish NixOS users definitely need this right now regarding all the bans and slowdowns recently.

**Review by Nishimara — APPROVED** [review]
this looks cuter than mine PR :D

**Review by Vagahbond — APPROVED** [review]

**Review by s0me1newithhand7s — APPROVED** [review]
tysm 🙏🏻for this, and welcome to the nixos, friend. looks good to me, tho, i'll add labels in case there missing some.

**Review by purrpurrn — APPROVED** [review]

File: nixos/modules/services/networking/zapret.nix
Comment (SigmaSquadron) [inline nixos/modules/services/networking/zapret.nix:115]: I'm assuming the associated package doesn't build a systemd unit?

File: nixos/modules/services/networking/zapret.nix
Comment (SigmaSquadron) [inline nixos/modules/services/networking/zapret.nix:116]:
```suggestion
          description = "DPI bypass service";
```
systemd units usually lack a period.

**Review by SigmaSquadron — COMMENTED** [review]
This is excellent! Aside from the comments below, the commits should be squashed according to the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).

Something like this should work:
1. `maintainers: add voronind`
2. `nixos/zapret: init`

**Comment by s0me1newithhand7s** [comment]
> This is excellent! Aside from the comments below, the commits should be squashed according to the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).
> 
> Something like this should work:
> 
> 1. `maintainers: add voronind`
> 2. `nixos/zapret: init`

fully agree with it.

**Comment by SigmaSquadron** [comment]
@OfBorg test simple

**Comment by ardishko** [comment]
> > This is excellent! Aside from the comments below, the commits should be squashed according to the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions).
> > Something like this should work:
> > 
> > 1. `maintainers: add voronind`
> > 2. `nixos/zapret: init`
> 
> fully agree with it.

Me too.

File: nixos/modules/services/networking/zapret.nix
Comment (voronind-com) [reply nixos/modules/services/networking/zapret.nix:30]: Yeah, I've already fixed that after full local testing. Weird that you don't see that.

File: nixos/modules/services/networking/zapret.nix
Comment (voronind-com) [reply nixos/modules/services/networking/zapret.nix:149]: Yes. Otherwise no traffic would be routed via service.

File: nixos/modules/services/networking/zapret.nix
Comment (SigmaSquadron) [reply nixos/modules/services/networking/zapret.nix:30]: You haven't pushed your changes :)

File: nixos/modules/services/networking/zapret.nix
Comment (voronind-com) [reply nixos/modules/services/networking/zapret.nix:115]: Correct. And no, no upstream unit.

**Comment by voronind-com** [comment]
@SigmaSquadron @ardishko Fixed comments and squashed.

**Comment by voronind-com** [comment]
> the quality of the module may vary

As I've already stated, this is my first time contribution. I'd really appreciate your suggestions!

**Comment by s0me1newithhand7s** [comment]
> @SigmaSquadron @ardishko Fixed comments and squashed.

now you can ask for re-review,i'll fix labels.

**Review by s0me1newithhand7s — APPROVED** [review]

**Comment by SigmaSquadron** [comment]
> As I've already stated, this is my first time contribution. I'd really appreciate your suggestions!

Rest assured that this is already a really good diff, so there's not a lot of changes to make. I'd like some eyes from more experienced reviewers in the firewall stuff though.

> Turkish & Russian users can already use this PR/MR, but the quality of the module may vary.

@purrpurrn Did something specific in this module appear broken or low-quality from your testing?

**Comment by voronind-com** [comment]
> > @purrpurrn Did something specific in this module appear broken or low-quality from your testing?
> 
> I wanted to point out Commit Conventions, but I was too slow to type it out. Otherwise I just wrote that because an open PR/MR that is still in review may have some unpredictable issues. So far it does look good though.

Looks like I've implemented everything that people asked. Asked for re-review.

**Comment by SigmaSquadron** [comment]
> Looks like I've implemented everything that people asked. Asked for re-review.

The type issue is still there, no?

**Comment by voronind-com** [comment]
> The type issue is still there, no?

I've replied to you and saw no response. The type is supposed to be like that because you have to specify parameters, and they can't have a default value.

**Review by TheMaxMur — APPROVED** [review]

**Comment by SigmaSquadron** [comment]
The default cannot be a value not allowed by the `type`. If the user needs to input some value, I recommend setting the default to an empty list (`[ ]`) and adding an evaluation warning if the option hasn't been changed by the user.

**Comment by voronind-com** [comment]
> The default cannot be a value not allowed by the `type`. If the user needs to input some value, I recommend setting the default to an empty list (`[ ]`) and adding an evaluation warning if the option hasn't been changed by the user.

One moment, I'll replace it with an assertion check.

**Comment by voronind-com** [comment]
Replaced the null type with `(builtins.length cfg.params) != 0` assertion check.

**Comment by SigmaSquadron** [comment]
Also, thanks for mentioning that other modules have similar requirements; a quick `rg` search showed a type bug in the moosefs module, which will need to be fixed eventually.

**Review by SigmaSquadron — APPROVED** [review]
LGTM, thank you for contributing to Nixpkgs!

**Comment by SigmaSquadron** [comment]
@OfBorg test simple

**Review by ardishko — APPROVED** [review]
I trust that that the default value issue will be dealt with soon, so I'm approving. It is very late where I am so I wanted to review and approve before I go to sleep.

The commits have been squashed accordingly, and now everything looks clean. LGTM.

**Comment by ardishko** [comment]
(my approval comment went through a little late, so ignore that first part)

**Review by purrpurrn — APPROVED** [review]

**Comment by voronind-com** [comment]
> LGTM, thank you for contributing to Nixpkgs!

<sup><sub>Sorry for the off-topic, but I'm a happy man now. I've been doing working as a full-time developer for 8 years now, and used Linux for about 10. I've always wanted to contribute to OSS but delayed it every time. Glad my first experience is so fantastic, all thanks to everyone here. Hope y'all have a great day.</sup></sub>

**Comment by s0me1newithhand7s** [comment]
ofborg test went well, yay!

**Comment by voronind-com** [comment]
I also re-tested the final version, works fine.

**Comment by voronind-com** [comment]
We just wait for the checks? I mean do I need to do anything else or just wait?

**Comment by TheSaintDiratof** [comment]
Where are nftables configuration and tpws? That's not all zapret, only nfqws+iptables config.

**Comment by voronind-com** [comment]
> Where are nftables configuration and tpws? That's not all zapret, only nfqws+iptables config.

Nftables should work as per nixos compatibility between them, no idea about tpws sadly. I was hoping on further improving this module when people actually ask for different implementations. For now, people with custom firewalls can set `configureFirewall = false` and do configuration themselves. And nfqws is what actually does the job, zapret binary is just a wrapper with configuration that sinked previous PR. This package actually works on NixOS systems. Hope.I answered your questions.

**Comment by TheSaintDiratof** [comment]
В таком случае я не вижу особых преимуществ вашего решения перед https://github.com/aca/zapret-flake.nix или решением nishimara.

**Comment by voronind-com** [comment]
> В таком случае я не вижу особых преимуществ вашего решения перед https://github.com/aca/zapret-flake.nix или решением nishimara.

Nishimara PR (original) had issues with rebuilding zapret package with configuration changes, which is obviously bad for binary caching reasons. This PR was made as an alternative that doesn't rebuild the package. And the flake you provided was not made into PR, so I can't comment on that.

**Comment by TheSaintDiratof** [comment]
At least rename services.zapret, please. That would help avoid name conflict with aca/zapret-flake.nix

**Comment by voronind-com** [comment]
> At least rename services.zapret, please. That would help avoid name conflict with aca/zapret-flake.nix

Sorry, I need other maintainers guidance on this. I don't see a reason why nixpkgs should follow flakes from the wild, but I might be wrong.

**Comment by SigmaSquadron** [comment]
> Sorry, I need other maintainers guidance on this. I don't see a reason why nixpkgs should follow flakes from the wild, but I might be wrong. For now it follows consistent logic of all other NixOS package-specific services. Anyway why would anyone use both at the same time?

There's no need to care about what other repos do. `services.zapret` is appropriate.

**Comment by Nishimara** [comment]
> In that case, I don't see much advantage of your solution over https://github.com/aca/zapret-flake.nix or nishimara's solution.

If you decide to keep this flake in inputs after merge, you need something like
```nix
disabledModules = [
  "services/networking/zapret.nix"
];
```

**Comment by ardishko** [comment]
Just dropping in to ask about OfBorg's checks since I'm not familiar. It's been about 12 hours, why havent the darwin checks been completed yet?

**Comment by SigmaSquadron** [comment]
> Just dropping in to ask about OfBorg's checks since I'm not familiar. It's been about 12 hours, why havent the darwin checks been completed yet?

Don't worry about it. Darwin checks are optional for Linux/NixOS changes and they usually take days to complete, especially near the final release staging cycles.

Emily has been working on improving the Darwin OfBorg CI situation; hopefully their completion times will approach Linux's soon.

**Comment by ardishko** [comment]
Can we skip the check if it's optional? I remember someone did that with a command once.

**Comment by SigmaSquadron** [comment]
No. I suggest ignoring it.

**Comment by voronind-com** [comment]
> No. I suggest ignoring it.

Then we still need someone who could merge this. Or is it automatic?

**Comment by SigmaSquadron** [comment]
It is not automatic. A committer will eventually merge this; please be patient.

**Comment by s0me1newithhand7s** [comment]
> > No. I suggest ignoring it.
> 
> Then we still need someone who could merge this. Or is it automatic?

that's funny how you got actually a lot approvals from members, but not from committers. 👀

**Comment by voronind-com** [comment]
Checks just passed. Darwin took like 3 days wow.

**Comment by s0me1newithhand7s** [comment]
> Checks just passed.

hooray!

**Comment by ardishko** [comment]
Hopefully a commiter sees this now.

**Review by kanashimia — APPROVED** [review]

**Comment by voronind-com** [comment]
I'm in chat with a commiter. We added the optionalString and now waiting for the ofborg eval before merging.

**Comment by s0me1newithhand7s** [comment]
> I'm in chat with a commiter. We added the optionalString and now waiting for the ofborg eval before merging. I also re-tested locally.

good luck, pal 👍🏻

**Review by s0me1newithhand7s — APPROVED** [review]

**Review by azahi — APPROVED** [review]

**Comment by s0me1newithhand7s** [comment]
congrats @voronind-com! and welcome to the nixos/nixpkgs. 🎉

**Comment by ardishko** [comment]
YOOOOOOOOOOOOOOOOOOOOOOOOOOO!!!!!! FINALLY!

**Pull request merged**
